site/content/blog/2021-04-23-optimizing-go-compiler-1.md

```diff
@@ -0,0 +1,227 @@
+
+Do not get discouraged for the length of the content: it does not correlate with the complexity of the issue, but with the level of detail I tried to use when explaining it. I wanted to make sure I explained the ins and outs of how I approached the fix, and I do not assume any knowledge about compilers from the reader, so I tried to discuss everything in chapter and verse.
+
+Enough with the intro, let's start!
+
+# I'd like to work on this one!
+
+Back in September 2020, my colleague [Agniva](https://github.com/agnivade) posted [a message in the Go channel](https://community.mattermost.com/core/pl/pd74uyf1z7rr9d46jfrww581na), where the Mattermost community hangs out to discuss all things Go, inviting people to start contributing to the compiler. He posted links to three issues in the Go project, one of which was [issue #41663](https://github.com/golang/go/issues/41663).
+
+Agniva told that this issue was "particularly suitable for newcomers looking to get started". That's me!, I said. I am a newcomer, and I am looking to get started. After double checking with him that "newcomer" meant someone that knew nothing about compilers, I picked it up, which was as easy as writing "[I'd like to work on this one!](https://github.com/golang/go/issues/41663#issuecomment-699878893)" on the issue itself.
+
+So that was it. I had then committed to fix an issue I knew nothing about. Of course, I started regretting that decision the very moment after I hit Send. I know now that it was a great decision, but at that time it looked quite daunting.
+
+# The issue
+
+Ok, so let's dive into the issue itself. The bug in the compiler can be explained with the following function:
+
+```go
+import "encoding/binary"
+
+func f(b []byte, x *[8]byte) {
+	_ = b[8]
+	t := binary.LittleEndian.Uint64(x[:])
+	binary.LittleEndian.PutUint64(b, t)
+}
+```
+
+The report said that this function compiled to a series of small `MOVx`, but that it could be optimized to use instead a couple of `MOVQ`: one for loading and one for storing.
+
+If this sounds as gibberish, that's ok. Bear with me and we will get to what that sentence means. First, let's see what the function is supposed to do. To make things a bit easier, let's simply rewrite the variable names and add a couple of comments:
+
+```go
+import "encoding/binary"
+
+func copyArrayToSlice(dest []byte, src *[8]byte) {
+	_ = dest[8]                                // 1. bounds check
+	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
+	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
+}
+```
+
+It is now more clear what the function does. It stores the contents of the second argument, the `src` array, into the first argument, the slice `dest`. It uses the [encoding/binary](https://golang.org/pkg/encoding/binary/) package for reading and writing the 8 bytes as an unsigned integer of 64 bits. For doing the copy, we use a temporal variable, `temp`.
```

Comment:
maybe "temporary" variable is better. "temporal" sounds like a sci-fi term :P

Reply:
Good catch, that's a mistake I always make!